ratelimits/source_redis.go

```diff
@@ -122,27 +142,38 @@ func (r *RedisSource) BatchGet(ctx context.Context, bucketKeys []string) (map[st
 	}
 	results, err := pipeline.Exec(ctx)
 	if err != nil {
-		r.latency.With(prometheus.Labels{"call": "batchget", "result": resultForError(err)}).Observe(time.Since(start).Seconds())
 		if !errors.Is(err, redis.Nil) {
+			r.observeLatency("batchget", r.clk.Since(start), err)
 			return nil, err
 		}
+		// One of the keys does not exist. We'll account for this when
+		// processing the results.
 	}
 
+	totalLatency := r.clk.Since(start)
+	perEntryLatency := totalLatency / time.Duration(len(bucketKeys))
+
+	var batchErr error
 	tats := make(map[string]time.Time, len(bucketKeys))
 	for i, result := range results {
 		tatNano, err := result.(*redis.StringCmd).Int64()
 		if err != nil {
-			if errors.Is(err, redis.Nil) {
-				// Bucket key does not exist.
-				continue
+			if !errors.Is(err, redis.Nil) {
+				// This should never happen as any errors should have been
+				// caught after the pipeline.Exec() call.
+				r.observeLatency("batchget", r.clk.Since(start), err)
+				return nil, err
 			}
-			r.latency.With(prometheus.Labels{"call": "batchget", "result": resultForError(err)}).Observe(time.Since(start).Seconds())
-			return nil, err
+			// Bucket key does not exist.
+			r.observeLatency("batchget_entry", perEntryLatency, err)
+			batchErr = errMixedSuccess
```

Comment:
I think this is not strictly true -- if _every_ result was a `redis.Nil` error, then the whole batch failed, but we're still going to record mixed success.